packages/gator-permissions-snap/src/ui/components/TokenIcon.tsx

```diff
@@ -8,17 +13,121 @@ export type TokenIconParams = {
   height?: number;
 };
 
-export const TokenIcon: SnapComponent<TokenIconParams> = ({
-  imageDataBase64,
-  altText,
-  width = 24,
-  height = 24,
-}) => {
-  if (!imageDataBase64) {
+const MAX_ICON_SIZE = 512;
+
+// Dangerous SVG elements and attributes that could execute scripts
+const DANGEROUS_SVG_ELEMENTS = [
+  'script',
+  'foreignObject',
+  'iframe',
+  'object',
+  'embed',
+  'link',
+  'style',
+];
+
+const DANGEROUS_SVG_ATTRIBUTES = [
+  'onload',
+  'onerror',
+  'onclick',
+  'onmouseover',
+  'onmouseout',
+  'onmousedown',
+  'onmouseup',
+  'onfocus',
+  'onblur',
+  'onchange',
+  'onsubmit',
+  'onreset',
+  'onselect',
+  'onkeydown',
+  'onkeypress',
+  'onkeyup',
+  'onabort',
+  'oncanplay',
+  'oncanplaythrough',
+  'ondurationchange',
+  'onemptied',
+  'onended',
+  'onloadeddata',
+  'onloadedmetadata',
+  'onloadstart',
+  'onpause',
+  'onplay',
+  'onplaying',
+  'onprogress',
+  'onratechange',
+  'onseeked',
+  'onseeking',
+  'onstalled',
+  'onsuspend',
+  'ontimeupdate',
+  'onvolumechange',
+  'onwaiting',
+  'href',
+  'xlink:href',
+];
+
+// zod schema for runtime validation
+const TokenIconParamsSchema = z.object({
+  imageDataBase64: z
+    .string()
+    .regex(
+      /^data:image\/(svg\+xml|png|jpeg|jpg);base64,[A-Za-z0-9+/=]+$/u,
+      'Must be a valid base64 data URI',
+    )
+    .refine((dataUri) => {
+      // Only validate SVG content for security
+      if (!dataUri.startsWith('data:image/svg+xml;base64,')) {
```

Comment:
I think the presumption is that the data _is_ base64 encoded png.

If the data doesn't startWith the expected prefix should we reject instead?

Reply:
Good Point!